core/rawdb/chain_iterator.go

```diff
@@ -0,0 +1,227 @@
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package rawdb
+
+import (
+	"errors"
+	"math"
+	"runtime"
+	"sync/atomic"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/prque"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+type (
+	prepareCallback func(*types.Block)              // Callback for custom concurrent pre-computations on block data
+	actionCallback  func(ethdb.Batch, *types.Block) // Callback for custom sequential operations on block data.
+)
+
+// iterateCanonicalChain iterates the specified range of canonical blocks and applies
+// the given action callback. Note, both for forward and backward iteration, the range
+// is [from, to).
+func iterateCanonicalChain(db ethdb.Database, from uint64, to uint64, prepare prepareCallback, action actionCallback, reverse bool, progMsg, doneMsg string) error {
+	// Short circuit if the action is nil or if the range is invalid
+	if action == nil {
+		return nil
+	}
+	if from >= to {
+		return nil
+	}
+	// Spawn multi-routines, iterate over the specified blocks and invoke prepare
+	// callback concurrently.
+	var (
+		number  int64
+		results = make(chan *types.Block, 4*runtime.NumCPU())
+	)
+	if !reverse {
+		number = int64(from - 1)
+	} else {
+		number = int64(to)
+	}
+	abort := make(chan struct{})
+	defer close(abort)
+
+	threads := to - from
+	if cpus := runtime.NumCPU(); threads > uint64(cpus) {
+		threads = uint64(cpus)
+	}
+	for i := 0; i < int(threads); i++ {
+		go func() {
+			for {
+				// Fetch the next task number, terminating if everything's done
+				var n int64
+				if !reverse {
+					n = atomic.AddInt64(&number, 1)
+					if n >= int64(to) {
+						return
+					}
+				} else {
+					n = atomic.AddInt64(&number, -1)
+					if n < int64(from) {
+						return
+					}
+				}
+				block := ReadBlock(db, ReadCanonicalHash(db, uint64(n)), uint64(n))
+				if prepare != nil && block != nil {
+					prepare(block)
+				}
+				// Feed the block to the aggregator, or abort on interrupt
+				select {
+				case results <- block:
+				case <-abort:
+					return
+				}
+			}
+		}()
+	}
+	// Reassemble the blocks into a contiguous stream and apply the action callback
+	var (
+		next, first, last int64
+		queue             = prque.New(nil)
+
+		batch  = db.NewBatch()
+		start  = time.Now()
+		logged = start.Add(-7 * time.Second) // Unindex during import is fast, don't double log
+	)
+	if !reverse {
+		next, first, last = int64(from), int64(from), int64(to)
+	} else {
+		next, first, last = int64(to-1), int64(to-1), int64(from-1)
+	}
+	for i := from; i < to; i++ {
+		// Retrieve the next result and bail if it's nil
+		block := <-results
+		if block == nil {
+			return errors.New("broken database")
+		}
+		// Push the block into the import queue and process contiguous ranges
+		priority := -int64(block.NumberU64())
```

Comment:
In our particular case with tx indexing, we don't actually require the blocks to come in a strict sequence, so could just operate on them as they come in. So we wouldn't have to mess about with the prque. But the que is fairly small, so maybe that's nothing to worry about

Actually, now that I think about it, does it really matter in `InitDatabaseFromFreezer` that we deal with the blocks strictly in order?

Reply:
If batches are atomic, then it doesn't matter.

Reply:
Ah no wait, it matters because we commit batches after some data limit is reached. If we don't push blocks into batches in an ordered way and terminate/crash geth in between, we might end up with gaps.

That said, we're not setting the progress marker anywhere, so it might all just be "lost" and recomputed on restart.

Reply:
Yes, strict order is required for many cases. Like the tx indexing, we always assume that the block above the `TAIL` has indexes available.

Reply:
Yes, strict order is required for many cases. Like the tx indexing, we always assume that the block above the `TAIL` has indexes available.